codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -498,15 +593,70 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx,
 
             // Encode the entire header block.
             headerBlock = ctx.alloc().buffer();
-            headersEncoder.encodeHeaders(streamId, headers, headerBlock);
+            try {
+                headersEncoder.encodeHeaders(streamId, headers, headerBlock);
+            } catch (Http2Exception e) {
+                return promise.setFailure(e);
+            }
 
             Http2Flags flags =
                     new Http2Flags().endOfStream(endStream).priorityPresent(hasPriority).paddingPresent(padding > 0);
-
-            // Read the first fragment (possibly everything).
             int nonFragmentBytes = padding + flags.getNumPriorityBytes();
             int maxFragmentLength = maxFrameSize - nonFragmentBytes;
-            ByteBuf fragment = headerBlock.readRetainedSlice(min(headerBlock.readableBytes(), maxFragmentLength));
+            int fragmentBytes = min(headerBlock.readableBytes(), maxFragmentLength);
```

Comment:
Here, instead, I've chosen to inline the fast-path and delegate to a separate method the slowest one, mostly because It seems way more common to hit it (binary headers size below <= 128 bytes seems pretty common to me).